projecttacoma/fqm-execution
# AsyncPatientSource Functionality and CLI Overhaul

# Summary
fqm-execution now supports calculation using a cql-exec-fhir AsyncPatientSource. Also cleaned up the CLI and changed the -o flag settings to commands

## New behavior
All existing functionality should be unchanged. Added the ability to pass in a fhir server url and a list of patient ids, to replace patient bundles as a calculation option. Uses cql-execution's new async behavior and cql-exec-fhir's new AsyncPatientSource and AsyncPatient to accomplish this.

## Code changes
### New cli options: 
- --fhir-server-url: a url to a fhir server that contains data for calculation
- --patient-ids: a list of ids which signifies which patients a measure will be run on when using an async patient source
### Updated cli options: 
-d, --debug -> --debug
-f, --as-patient-source -> --as-patient-source
-c, --cache-valuesets -> --cache-valuesets
-a, --vs-api-key -> --vs-api-key
-r, --report-type -> --report-type
### Other changes
Updated Execution to properly await cql-execution execute function
Updated gaps code to correctly await expression executions
Updating testing to account for new async functions

# Testing guidance
- Run unit tests
- Make sure you have the most recent versions of projecttacoma/cql-execution, projecttacoma/cql-exec-fhir, and projecttacoma/deqm-test-server
- Reset your test server db using `npm run db:reset`, then `npm run connectathon-upload`
- Link cql-exec-fhir to fqm-execution by running `npm link path/to/projecttacoma/cql-exec-fhir`
- Link cql-execution to fqm-execution by running `npm link path/to/projectacoma/cql-execution`
- Start up deqm-test-server using `npm start`
- Run `npm run build` in fqm-execution
- Ping me for a file of requests to use for testing. The async patient source version of each request should match the same request when run with regular patient bundles
- Run your own test commands! Make sure you run through each output type and report type.
- Test out invalid combinations of options and make sure that a proper error is thrown
- Look over the cli and the readme to make sure everything is documented correctly. Try running `node build/cli.js -h` and make sure the help command is useful and properly formatted.
